### ISSUE TITLE
Poor performance on a simple dataset

### ISSUE THREAD

**Issue by Kodiologist** [ja-thomas/autoxgboost#62]
I'm not familiar with the statistical approach taken by mlrMBO, so excuse me if I'm missing something. Anyway, I was going to ask a question about overfitting in autoxgboost, but it looks like I actually have an underfitting problem. Below is a simple example using DART and mostly default settings. Training and test error for the autoxgboost model hovers near the SD and is much worse than that of linear regression. What am I doing wrong?

    library(autoxgboost)
    library(mlr)
    library(ParamHelpers)

    set.seed(456)
    xgb.threads = 10

    autoxgbparset.dart = makeParamSet(
        makeNumericParam("eta", lower = 0.01, upper = 0.2),
        makeNumericParam("gamma", lower = -7, upper = 6, trafo = function(x) 2^x),
        makeIntegerParam("max_depth", lower = 3, upper = 20),
        makeNumericParam("colsample_bytree", lower = 0.5, upper = 1),
        makeNumericParam("colsample_bylevel", lower = 0.5, upper = 1),
        makeNumericParam("lambda", lower = -10, upper = 10, trafo = function(x) 2^x),
        makeNumericParam("alpha", lower = -10, upper = 10, trafo = function(x) 2^x),
        makeNumericParam("subsample", lower = 0.5, upper = 1),
        makeDiscreteParam("booster", values = "dart"),
        makeDiscreteParam("sample_type", values = c("uniform", "weighted")),
        makeDiscreteParam("normalize_type", values = c("tree", "forest")),
        makeNumericParam("rate_drop", lower = 0, upper = 1),
        makeNumericParam("skip_drop", lower = 0, upper = 1),
        makeLogicalParam("one_drop"))

    N = 2000
    d = transform(data.frame(
        x1 = rnorm(N),
        x2 = rnorm(N),
        x3 = rnorm(N)),
        y = 2*x2 + (abs(x3) < 1) + rnorm(N))
    train = (1 : N) <= 1000

    m.lm = lm(y ~ ., data = d[train,])
    m.axgb = autoxgboost(
        task = makeRegrTask(target = "y", data = d[train,]),
        measure = rmse,
        par.set = autoxgbparset.dart,
        design.size = 30L,
        nthread = xgb.threads)

    f = function(a, b) sqrt(mean((a - b)^2))
    print(t(data.frame(
        SD = sd(d[!train, "y"]),
        perfect.train = f(d[train, "y"],
            with(d[train,], 2*x2 + (abs(x3) < 1))),
        perfect.test = f(d[!train, "y"],
            with(d[!train,], 2*x2 + (abs(x3) < 1))),
        lm.train = f(d[train, "y"], predict(m.lm)),
        lm.test = f(d[!train, "y"], predict(m.lm, newdata = d[!train,])),
        axgb.train = f(d[train, "y"],
            predict(m.axgb, newdata = d[train,])$data$response),
        axgb.test = f(d[!train, "y"],
            predict(m.axgb, newdata = d[!train,])$data$response))))

Output:

                      [,1]
    SD            2.258125
    perfect.train 1.001122
    perfect.test  1.021544
    lm.train      1.119062
    lm.test       1.137798
    axgb.train    2.114853
    axgb.test     2.234926

CCing my cow-orkers: @allanjust, @liuyanguu

**Comment by liuyanguu** [comment]
I had an interesting further finding. The problem might lie in the predict function of autoxgboost. If I extract the parameters using `mlr::getHyperPars` and run a separate `xgboost::xgboost`, both the testing and training error went back to around 1.1, which looks right. I am glad I always extracted the hyperparameters... , Strangely enough, I have never identified such a problem before...

**Comment by liuyanguu** [comment]
@ja-thomas , I am not very familiar with mlr... I am curious how is the `predict` function called when the object is autoxgboost?

**Comment by ja-thomas** [comment]
Hi,

sorry for the late reply, I was away for a few days.

thanks for the issue, this is indeed very surprising and I found the problem to be that we call `cpoDropConstants` which seems to drop features that are far away from constant. This is a bug in mlrCPO.

For now I'll drop this step from the preprocessing, until it is fixed in mlrCPO.
